### PR TITLE
Use SonarIssue instead of tslint.RuleFailure

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: sonarts-core/src/utils/sonar-analysis.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+      primaryLocation.message!,
+      ruleName,
+    );
+
+    this.primaryLocation = primaryLocation;
+  }
+
+  public toJson() {
+    return {
+      failure: this.primaryLocation.message!,
```
Comment (SonarTech) [inline sonarts-core/src/utils/sonar-analysis.ts]: Bug Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=sonar-issue&issues=AWCdkH8rqDrCo5ysf9j1&open=AWCdkH8rqDrCo5ysf9j1)

[ISSUE_KEY]: AWCdkH8rqDrCo5ysf9j1

File: sonarts-core/src/utils/sonar-analysis.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  private cost?: number;
+  public readonly primaryLocation: IssueLocation;
+  private secondaryLocations: IssueLocation[] = [];
+
+  public constructor(primaryLocation: IssueLocation, ruleName: string) {
+    super(
+      primaryLocation.node.getSourceFile(),
+      primaryLocation.node.getStart(),
+      primaryLocation.lastNode.getEnd(),
+      primaryLocation.message!,
```
Comment (SonarTech) [inline sonarts-core/src/utils/sonar-analysis.ts]: Bug Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=sonar-issue&issues=AWCdkH8rqDrCo5ysf9j0&open=AWCdkH8rqDrCo5ysf9j0)

[ISSUE_KEY]: AWCdkH8rqDrCo5ysf9j0

File: sonarts-core/src/rules/mccabeComplexityRule.ts [hunk 3]
```diff
@@ -59,17 +59,15 @@ class Visitor extends SonarRuleVisitor {
...
-    if (is(node, ...FUNCTION_LIKE)) {
-      const functionComplexity = getFunctionComplexityNodes(node as ts.FunctionLikeDeclaration).length;
-      if (functionComplexity > this.threshold) {
-        this.addIssue(
-          functionLikeMainToken(node as ts.FunctionLikeDeclaration),
-          Rule.message(functionComplexity, this.threshold),
-        );
-      }
+  public visitFunctionLikeDeclaration(node: ts.FunctionLikeDeclaration) {
+    const functionComplexity = getFunctionComplexityNodes(node as ts.FunctionLikeDeclaration).length;
```
Comment (stas-vilchik) [inline sonarts-core/src/rules/mccabeComplexityRule.ts]: remove type cast

File: sonarts-core/src/rules/mccabeComplexityRule.ts [hunk 4]
```diff
@@ -59,17 +59,15 @@ class Visitor extends SonarRuleVisitor {
...
+    if (functionComplexity > this.threshold) {
+      this.addIssue(
+        functionLikeMainToken(node as ts.FunctionLikeDeclaration),
```
Comment (stas-vilchik) [inline sonarts-core/src/rules/mccabeComplexityRule.ts]: remove type cast

File: sonarts-core/src/rules/noRedundantParenthesesRule.ts [hunk 5]
```diff
@@ -42,13 +42,12 @@ export class Rule extends Lint.Rules.AbstractRule {
...
   private static readonly MESSAGE = "Remove these useless parentheses.";
 
-  visitNode(node: ts.Node): void {
-    if (is(node, ts.SyntaxKind.ParenthesizedExpression)) {
-      const parenthesizedExpression = node as ts.ParenthesizedExpression;
-      if (is(parenthesizedExpression.expression, ts.SyntaxKind.ParenthesizedExpression)) {
-        this.addIssue(node, Visitor.MESSAGE);
-      }
+  public visitParenthesizedExpression(node: ts.ParenthesizedExpression) {
+    const parenthesizedExpression = node as ts.ParenthesizedExpression;
```
Comment (stas-vilchik) [inline sonarts-core/src/rules/noRedundantParenthesesRule.ts]: remove type cast

File: sonarts-core/src/rules/mccabeComplexityRule.ts [hunk 6]
```diff
@@ -49,29 +50,24 @@ export class Rule extends tslint.Rules.AbstractRule {
...
-        this.addFailureAtNode(
-          functionLikeMainToken(node as ts.FunctionLikeDeclaration),
-          Rule.message(functionComplexity, this.threshold),
-        );
-      }
+  public visitFunctionLikeDeclaration(node: ts.FunctionLikeDeclaration) {
+    const functionComplexity = getFunctionComplexityNodes(node as ts.FunctionLikeDeclaration).length;
+    if (functionComplexity > this.threshold) {
+      this.addIssue(
+        functionLikeMainToken(node as ts.FunctionLikeDeclaration),
```
Comment (SonarTech) [inline sonarts-core/src/rules/mccabeComplexityRule.ts]: Code Smell Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS4325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=sonar-issue&issues=AWCd633F_Q-0dz4JAbfN&open=AWCd633F_Q-0dz4JAbfN)

[ISSUE_KEY]: AWCd633F_Q-0dz4JAbfN

File: sonarts-core/src/rules/noRedundantParenthesesRule.ts [hunk 7]
```diff
@@ -34,20 +35,19 @@ export class Rule extends Lint.Rules.AbstractRule {
...
   private static readonly MESSAGE = "Remove these useless parentheses.";
 
-  visitNode(node: ts.Node): void {
-    if (is(node, ts.SyntaxKind.ParenthesizedExpression)) {
-      const parenthesizedExpression = node as ts.ParenthesizedExpression;
-      if (is(parenthesizedExpression.expression, ts.SyntaxKind.ParenthesizedExpression)) {
-        this.addFailureAtNode(node, Walker.MESSAGE);
-      }
+  public visitParenthesizedExpression(node: ts.ParenthesizedExpression) {
+    const parenthesizedExpression = node as ts.ParenthesizedExpression;
```
Comment (SonarTech) [inline sonarts-core/src/rules/noRedundantParenthesesRule.ts]: Code Smell Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS4325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=sonar-issue&issues=AWCd631y_Q-0dz4JAbfK&open=AWCd631y_Q-0dz4JAbfK)

[ISSUE_KEY]: AWCd631y_Q-0dz4JAbfK

File: sonarts-core/src/rules/noIdenticalFunctionsRule.ts [hunk 8]
```diff
@@ -89,18 +90,16 @@ export class Rule extends tslint.Rules.AbstractRule {
...
-  protected visitNode(node: ts.Node): void {
-    if (is(node, ...FUNCTION_LIKE)) {
-      const body = (node as ts.FunctionLikeDeclaration).body;
-      if (is(body, ts.SyntaxKind.Block) && Walker.isBigEnough(body as ts.Block)) {
-        this.functionBlocks.push(body as ts.Block);
-      }
+  public visitFunctionLikeDeclaration(node: ts.FunctionLikeDeclaration) {
+    const { body } = node;
+    if (body && ts.isBlock(body) && Visitor.isBigEnough(body)) {
+      this.functionBlocks.push(body as ts.Block);
```
Comment (SonarTech) [inline sonarts-core/src/rules/noIdenticalFunctionsRule.ts]: Code Smell Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS4325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=sonar-issue&issues=AWCd632l_Q-0dz4JAbfL&open=AWCd632l_Q-0dz4JAbfL)

[ISSUE_KEY]: AWCd632l_Q-0dz4JAbfL

File: sonarts-core/src/rules/mccabeComplexityRule.ts [hunk 9]
```diff
@@ -49,29 +50,24 @@ export class Rule extends tslint.Rules.AbstractRule {
...
-    if (is(node, ...FUNCTION_LIKE)) {
-      const functionComplexity = getFunctionComplexityNodes(node as ts.FunctionLikeDeclaration).length;
-      if (functionComplexity > this.threshold) {
-        this.addFailureAtNode(
-          functionLikeMainToken(node as ts.FunctionLikeDeclaration),
-          Rule.message(functionComplexity, this.threshold),
-        );
-      }
+  public visitFunctionLikeDeclaration(node: ts.FunctionLikeDeclaration) {
+    const functionComplexity = getFunctionComplexityNodes(node as ts.FunctionLikeDeclaration).length;
```
Comment (SonarTech) [inline sonarts-core/src/rules/mccabeComplexityRule.ts]: Code Smell Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS4325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=sonar-issue&issues=AWCd633F_Q-0dz4JAbfM&open=AWCd633F_Q-0dz4JAbfM)

[ISSUE_KEY]: AWCd633F_Q-0dz4JAbfM

File: sonarts-core/src/utils/sonar-analysis.ts (hunk 1 above)
Comment (SonarTech) [reply sonarts-core/src/utils/sonar-analysis.ts]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: sonarts-core/src/utils/sonar-analysis.ts (hunk 2 above)
Comment (SonarTech) [reply sonarts-core/src/utils/sonar-analysis.ts]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: sonarts-core/src/rules/noRedundantParenthesesRule.ts [hunk 10]
```diff
@@ -43,8 +43,7 @@ class Visitor extends SonarRuleVisitor {
   private static readonly MESSAGE = "Remove these useless parentheses.";
 
   public visitParenthesizedExpression(node: ts.ParenthesizedExpression) {
-    const parenthesizedExpression = node as ts.ParenthesizedExpression;
-    if (is(parenthesizedExpression.expression, ts.SyntaxKind.ParenthesizedExpression)) {
+    if (is(node.expression, ts.SyntaxKind.ParenthesizedExpression)) {
```
Comment (stas-vilchik) [inline sonarts-core/src/rules/noRedundantParenthesesRule.ts]: super minor: I'd use `ts.isParenthesizedExpression`

**Pull request merged**